[feature] Added Object Notification Setting #38

pandafy opened the pull request:
- Added ObjectNotification model to add option to disable
  notification for a particular device
- Added API endpoints for ObjectNotification model

Closes #38

nemesisdesign commented on openwisp_notifications/base/models.py:
```diff
@@ -250,3 +251,15 @@ def web_notification(self):
         return self.type_config.get('web_notification')
+
+
+class AbstractObjectNotification(UUIDModel):
```
from reading this code is not immediately clear what this model does.

Does it serve as ignore?
Does every object need this? Or is it added only when a user deactivated notifications for a particular object?

pandafy commented:
I have renamed `ObjectNotification` to `IgnoreObjectNotification`, I hope it is more descriptive. I have also added doc string to the abstract model to make it explicit what this model is for. 

I have used `button` elements for the drop down widget and tested accessibility from keyboard.
![Hnet-image (20)](https://user-images.githubusercontent.com/32094433/91313273-d3c0d580-e7d2-11ea-8637-1e97ca462a4d.gif)

> ```css
> .ow-object-notification-option-container p {
>     margin: auto;
>     font-size: 12px;
>     padding: 8px 0;
>     outline: 0.5px solid rgba(0, 0, 0, 0.05);
> }
> ```

Instead of outline here, i went with border. While focusing though buttons, an outline appears which is of a darker colour(`#101010`) inherited from`user agent stylesheet` . If we overide this setting, it makes it almost impossible to see which element is in focus.  

> I clicked on Unsubscribe, the tooltip appearing on hover to the button has stayed "You are receiving notifications for this object."
Fixed it. :sweat_smile: 
![Hnet-image (22)](https://user-images.githubusercontent.com/32094433/91316269-40899f00-e7d6-11ea-922d-53a881058e75.gif)

pandafy commented:
That was a silly mistake. I have updated the code. 
![Screenshot_20200827_185730](https://user-images.githubusercontent.com/32094433/91453587-9117ff80-e89d-11ea-94aa-29eedcf6af2e.png)

nemesisdesign commented on README.rst:
```diff
@@ -575,6 +575,17 @@ by disabling both web and email option for a notification setting.
 then ``email_notification`` or ``web_notification`` option of concerned notification type will be used
 respectively.
 
+
```
let's avoid this extra blank line for consistency

nemesisdesign commented on openwisp_notifications/widgets.py:
```diff
@@ -0,0 +1,39 @@
+    css = {'all': ['openwisp-notifications/css/object-notifications.css']}
+
+
+def add_object_notification_widget():
```
this function is a bit magical. I would flag it as private (rename it to `_add_object_notification_widget`) so we can change it without worrying too much about backward compatibility in the future if we need it

nemesisdesign commented on openwisp_notifications/widgets.py:
```diff
@@ -0,0 +1,39 @@
+        'admin/js/jquery.init.js',
+        'openwisp-notifications/js/object-notifications.js',
+    ]
+    css = {'all': ['openwisp-notifications/css/object-notifications.css']}
```
could you convert these two lists to tuples for consistency with the other modules please?

nemesisdesign commented on openwisp_notifications/tasks.py:
```diff
@@ -202,3 +215,11 @@ def ns_organization_created(instance_id):
+
+
+@shared_task
+def delete_ignore_object_notification_object(instance_id):
```
i'd rename it to `delete_ignore_object_notification`

nemesisdesign commented on openwisp_notifications/templates/admin/change_form.html:
```diff
@@ -0,0 +1,24 @@
+    {{ block.super }}
+    <script>
+        const owNotifyAppLabel = "{{ opts.app_label }}",
+            owNotifyModelName = "{{ opts.model_name }}",
```
here you're using double quotes but single quotes below, let's keep consistency and use single quotes

nemesisdesign commented on openwisp_notifications/templates/admin/change_form.html:
```diff
@@ -0,0 +1,24 @@
+{% extends 'admin/change_form.html' %}
+
+{% block admin_change_form_document_ready %}
+    {{ block.super }}
```
I think we need to add a comment (using the django template comment syntax `{# comment #} that explains what this template does)

nemesisdesign reviewed (changes requested):
I'm not sure what went wrong, but while testing this branch on the latest openwisp-monitoring and openwisp-controller, the button disappeared :sob: 

Can you test the same and let me know?

pandafy commented:
I removed the changes from openwisp-monitoring which brings this feature as per your suggestion to move it to openwisp-controller altogether. For testing, please add following in `settings.py`
```
OPENWISP_NOTIFICATIONS_OBJECT_NOTIFICATION_WIDGET = ['openwisp_controller.config.admin.Device']
```

I will configure this in openwisp-controller in https://github.com/openwisp/openwisp-controller/pull/265 (I am currently working on it.)

nemesisdesign commented on README.rst:
```diff
@@ -759,6 +769,26 @@ It sets the number of seconds the notification contents should be stored in the
 If you want cached notification content to never expire, then set it to ``None``.
 Set it to ``0`` if you don't want to store notification contents in cache at all.
 
+``OPENWISP_NOTIFICATIONS_OBJECT_NOTIFICATION_WIDGET``
```
The term "object notification" may not be super clear for a user, I'm thinking if we can make this more explicit... what about:

- `OPENWISP_NOTIFICATIONS_IGNORABLE_ADMIN_CLASSES`
- `OPENWISP_NOTIFICATIONS_IGNORABLE_OBJECTS`
- `OPENWISP_NOTIFICATIONS_IGNORE_ENABLED_ADMIN`

Or anything else along these lines? Any better idea?

pandafy replied:
`OPENWISP_NOTIFICATIONS_IGNORE_ENABLED_ADMIN` sounds good. :smile:

nemesisdesign commented on openwisp_notifications/checks.py:
```diff
@@ -27,3 +27,32 @@ def check_cors_configuration(app_configs, **kwargs):
+                obj='Settings',
+            )
+        )
+    for model_admin in app_settings.OW_OBJECT_NOTIFICATION_WIDGET:
```
model_admin is technically not an instance of `ModelAdmin` but a python `path`, while reading this code I thought it was checking if the path was an instance of `ModelAdmin` and got confused.

I think it would make the code more readable to rename `model_admin` to `path`.

nemesisdesign commented on openwisp_notifications/checks.py:
```diff
@@ -27,3 +27,32 @@ def check_cors_configuration(app_configs, **kwargs):
+                    ),
+                    obj='Settings',
+                )
+            )
```
then here we could import the class using [import_string](https://docs.djangoproject.com/en/3.1/ref/utils/#django.utils.module_loading.import_string) and ensure is an instance of `ModelAdmin` as well as mentioned above.

nemesisdesign reviewed (commented):
> I removed the changes from openwisp-monitoring which brings this feature as per your suggestion to move it to openwisp-controller altogether. For testing, please add following in `settings.py`
> 
> ```
> OPENWISP_NOTIFICATIONS_OBJECT_NOTIFICATION_WIDGET = ['openwisp_controller.config.admin.Device']
> ```

I tried setting it as per your suggestion but the button did not appear, then, after several minutes of debugging I realized it should be
`DeviceAdmin` instead of `Device`, which in turn made me think, why don't we perform a check on this?

We could raise a system check error if an element in the list is not an instance of django `ModelAdmin`, and advise a fix, which would make it easier for users to understand, what do you think?

The rest looks good!

pandafy commented:
> I tried setting it as per your suggestion but the button did not appear, then, after several minutes of debugging I realized it should be
> `DeviceAdmin` instead of `Device`, which in turn made me think, why don't we perform a check on this?

Apologies for misleading you. :sweat_smile: I have added the required checks.

@nemesisdesign did you notice that I had referred this feature as object notification in README when describing it to users. I tried referencing as "ignore object notification widget" but in my opinion it sounds a bit odd. Do you think this inconsistency is beneficial or will it only come haunting us in future?

nemesisdesign commented on README.rst:
```diff
@@ -586,6 +586,16 @@ by disabling both web and email option for a notification setting.
+Using the widget on an object's admin form, a user can disable notifications for that object
+for a day, week, month or even forever.
+
+**Note**: This feature requires configuring `"OPENWISP_NOTIFICATIONS_IGNORE_ENABLED_ADMIN" <#openwisp-notifications-object-notification-widget>`_
```
the internal reference seems mistaken, should be `openwisp_notifications_ignore_enabled_admin`

nemesisdesign reviewed (approved):
Thanks!

I found some details about the README to improve, will send improvements directly to dev shortly.
